std/hash/keccaf.asm

```diff
@@ -0,0 +1,142 @@
+use std::array;
+
+// TODO
+let<T1, T2> routine: T1, T1, (T1, T1 -> T1), (T1, T1 -> T1), (T1, T1 -> T1) -> T2 =
+    |a, b, xor, and, or| a + b;
+
+// function definition
+let or: int, int -> int = |a, b| a | b;
+let and: int, int -> int = |a, b| a & b;
+let xor: int, int -> int = |a, b| a ^ b;
+
+// ln 9 - 12
+// TODO: check that these are good for word size 32
+let RHO: int[24] = [
+    1,  3,  6,  10, 15, 21, 28, 36, 45, 55, 2,  14,
+    27, 41, 56, 8,  25, 43, 62, 18, 39, 61, 20, 44
+];
+
+// ln 14 - 16
+// TODO: check that these are good for word size 32
+let PI: int[24] = [
+    10, 7,  11, 17, 18, 3, 5,  16, 8,  21, 24, 4,
+    15, 23, 19, 13, 12, 2, 20, 14, 22, 9,  6,  1
+];
+
+// ln 19 - 28
+// TODO: check that these are good for word size 32
+let RC: int[24] = [
+    0x0000000000000001, 0x0000000000008082, 0x800000000000808a,
+    0x8000000080008000, 0x000000000000808b, 0x0000000080000001,
+    0x8000000080008081, 0x8000000000008009, 0x000000000000008a,
+    0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
+    0x000000008000808b, 0x800000000000008b, 0x8000000000008089,
+    0x8000000000008003, 0x8000000000008002, 0x8000000000000080,
+    0x000000000000800a, 0x800000008000000a, 0x8000000080008081,
+    0x8000000000008080, 0x0000000080000001, 0x8000000080008008
+];
+
+// ln 30 - 33
+// left rotation
+let rotl32: u32, u32 -> u32 = |x, n| or((x << n), (x >> (32 - n)));
+
+// ln 35 - 40
+// change endianness for a 32 bit number byte by byte
+// e.g. 0xaabbccdd -> 0xddccbbaa
+let swap_u32: int -> int = |x| 
+    or((or(and((x << 8), 0xff00ff00), and((x >> 8), 0x00ff00ff)) >> 16),
+    (or(and((x << 8), 0xff00ff00), and((x >> 8), 0x00ff00ff)) << 16)); 
+
+// compression function
+
+// ln 44
+let bc: int[5] = [0, 0, 0, 0, 0];
+
+// ln 45
+let t: int = 0;
+
+// ln 47 - 49
+let swap_u32_loop: int[25] -> int[25] = |st| array::map(array::new(25, |i| i), |i| swap_u32(st[i]));
+
+// ln 53 - 55
+let theta_bc: int[25], i -> int = |st, i| xor(st[i], xor(st[i + 5], xor(st[i + 10], xor(st[i + 15], st[i + 20]))));
+let theta_bc_loop: int[25] -> int[5] = |st| array::map(array::new(5, |i| i), |i| theta_bc(st, i));
+
+// ln 57 - 62
+let theta_t: int[5] -> int = |bc, i| xor(bc[(i + 4) % 5], rotl32(bc[(i + 1) % 5], 1));
+let theta_st: int[25], int, int, int -> int = |st, t, j, i| xor(st[(j * 5) + i], t);
+let theta_st_loop: int[25], int, int -> int[5] = |st, t, i| array::map(array::new(5, |i| i), |j| theta_st(st, t, j, i));
+let theta_t_loop: int[5], int[25], int -> int[5][5] = |bc, st, t| array::map(array::new(5, |i| i), |i| {
+    let t = theta_t(bc, i); // int
+    let st_partial = theta_st_loop(st, t, i); // int[5]
+    st_partial
+});
+// rearrange st returned in ln 60 - version 1: calls theta_t_loop
```

Comment:
I don't understand, why do you need to rearrange?

Once the 5 bc values are computed in line 54, we can compute the new st values in line 60 in the following way

```
let theta = |bc, st| std::array::map_enumerated(st, |k, x| {
  let i = k % 5;
  let j = k / 5;
  let t = bc[(i + 4) % 5] ^ rotl64(bc[(i + 1) % 5], 1);
  xor(x, t)
});
```

Reply:
Agree on this. I just didn't know this kind of syntax for nested loop but now I do!